Branch

anmivin opened the pull request:
полный проект)))

anmivin committed:
Create and first commit

anmivin committed:
хакер на костылях

anmivin committed:
made some changes

anmivin committed:
delete

Stilwuc commented on hackers-site/package.json:
```diff
@@ -0,0 +1,45 @@
+    "react-router-dom": "^6.3.0",
+    "react-scripts": "5.0.1",
+    "styled-components": "^5.3.5",
+    "swr": "^1.3.0",
```
swr не юзаешь

Stilwuc commented on hackers-site/src/App.js:
```diff
@@ -0,0 +1,71 @@
+              
+  
+
+</BrowserRouter> 
```
С включенным prettier на save ты меня обманула)
Вижу очень много косяков форматирования.
Поднастрой его и пройдись по всем файлам плз

anmivin replied:
не обманывала, оно правда на save стоит, только не работает( настрою

Stilwuc commented on hackers-site/src/components/New.js:
```diff
@@ -0,0 +1,211 @@
+ 
+    const goToNew = (key) => {
+        setContext(key);
+        console.log(key);
```
console log и debugger убирай перед комитом/ПР)

anmivin replied:
фиксед

Stilwuc commented on hackers-site/src/TheContext.js:
```diff
@@ -0,0 +1,5 @@
+import React, { createContext, useContext } from "react";
+
+const Context = createContext();
```
Слишком абстрактное название)
Делай более конкретные названия, например тут надо назвать как то чтобы было понятно за что отвечает твой context.
В твоем случае наверно что то типо `NewsItemContext`

anmivin replied:
фиксед

Stilwuc commented on hackers-site/src/App.js:
```diff
@@ -0,0 +1,71 @@
+text-align: center;
+
+`
+const Navi = styled.nav `
```
Почему именно `Navi` а не просто `Nav`?)
Если фанатка Na`Vi в ксочке так и скажи, отстану :D

anmivin replied:
не, я глор спиритов, и то в дотке

Stilwuc replied:
Тоже почетно я считаю)

Stilwuc commented on hackers-site/src/App.js:
```diff
@@ -0,0 +1,71 @@
+import Context from './TheContext';
+import {QueryClient, QueryClientProvider} from 'react-query';
+
+const TheHeader = styled.div `
```
Стараемся не юзать всякие артикли в названиях (проще читать)
Просто Header

Stilwuc commented on hackers-site/src/components/New.js:
```diff
@@ -0,0 +1,211 @@
+      );
+      
+
+      const { data: second} = useQuery(['newses'], 
```
У тебя 4 запроса useQuery и каждый из них с одним и тем же ключом.
Это больно, так как в итоге у тебя будет только одна пачка кэшироваться под этим ключом

anmivin replied:
фиксед

Stilwuc commented on hackers-site/src/components/New.js:
```diff
@@ -0,0 +1,211 @@
+    
+   
+  
+    const {isLoading, error, data: first, refetch} = useQuery(['newses'], 
```
Юзаешь только isLoading и error первого запроса
А что если сваляться другие?)
Все еще советую сделать 4 axios.get просто в одной функции, типо такого 
```
const { isLoading, error, data: first, refetch } = useQuery(['newses'], async () =>{
  const res1 = await axios.get('https://api.hnpwa.com/v0/newest/1.json');
  const res2 = await axios.get('https://api.hnpwa.com/v0/newest/2.json');
  const res3 = await axios.get('https://api.hnpwa.com/v0/newest/3.json');
  const res4 = await axios.get('https://api.hnpwa.com/v0/newest/4.json');
  return [...res1 .data, res2.data, res3.data, res4.data];
})
```
Вот, но еще лучше если ты сделаешь все эти 4 запроса в этой функции в паралель, с использованием `Promise.all`

anmivin replied:
фиксед

Stilwuc commented on hackers-site/src/components/New.js:
```diff
@@ -0,0 +1,211 @@
+    return (
+      <div>
+        
+        <TheButton onClick={() => refreshPage()}> Refresh </TheButton>
```
В данной ситуации можно передать функции как ссылку, чтобы меньше писать)
```suggestion
        <TheButton onClick={refreshPage}> Refresh </TheButton>
```

anmivin replied:
фиксед

Stilwuc commented on hackers-site/src/components/New.js:
```diff
@@ -0,0 +1,211 @@
+          
+                                  
+            {
+              first.map((newses) =>
```
Ну дублированный map 4 раза это конечно не очень, не одобряю
Стараемся не дублировать код, а то писать что то 4 раза вместо одного душно

anmivin replied:
фиксед

Stilwuc commented on hackers-site/src/components/New.js:
```diff
@@ -0,0 +1,211 @@
+                        )
+                    }
+                    {
+              fourth.map((newses) =>
```
У тебя 4 куска данных длинной по 30 элементов.
В итоге рендерится 120 элементов, а в ТЗ было 100)

anmivin replied:
фиксед

Stilwuc commented on hackers-site/src/components/New.js:
```diff
@@ -0,0 +1,211 @@
+                    {
+              second.map((newses) =>
+              
+                  <div key={newses.id} >
```
Эх а вот если бы разбила на компоненты было бы кайф

Stilwuc commented on hackers-site/src/components/NewData.js:
```diff
@@ -0,0 +1,104 @@
+function New() {
+    
+  const[context, setContext] = useContext(Context);
+  const param = context;
```
Я кажется понимаю откуда это взялось но это не нужно)
Просто убери на 46 строчки context и тогда никто не будет перезаписывать реальное значение context

Stilwuc commented on hackers-site/src/components/NewData.js:
```diff
@@ -0,0 +1,104 @@
+    
+  const {isLoading, error, data, refetch} = useQuery(['onenew'], 
+  async ({context}) => { 
+    const res = await axios.get(`https://api.hnpwa.com/v0/item/${param}.json`)
```
Помнишь я говорил что значение контекста живет до f5.
Как ты думаешь что будет когда я выберу новость и нажму f5?)
Надо фиксануть

Stilwuc commented on hackers-site/src/components/NewData.js:
```diff
@@ -0,0 +1,104 @@
+
+  const [isHidden, setIsHidden] = useState(true);
+
+  const Comments = ({ items }) => (
```
Объявлять компонент в компоненте очень плохо.
У тебя лютая деоптимизация будет, так как на каждый ререндер родителя будет заново создаваться компонент Comment (не перерендериваться а создаваться)

Stilwuc commented on hackers-site/src/components/NewData.js:
```diff
@@ -0,0 +1,104 @@
+
+  const Comments = ({ items }) => (
+    <React.Fragment>
+      {items.map(n => (
```
Переменные в одну букву смерть.
Что такое `n` вне контекста - вообще фиг знает

anmivin replied:
фиксед

Stilwuc commented on hackers-site/src/components/NewData.js:
```diff
@@ -0,0 +1,104 @@
+      {items.map(n => (
+        <div key={n.id}>
+          <h3>{n.user}</h3>
+          <div>{n.content.split('<p>').join(' ').split('&#x27;').join('`')}</div>
```
Интересная обработка)
А что если типы вставят туда ссылку `<a>` или лист `<li>`
Почитай про https://ru.reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml

Stilwuc commented on hackers-site/src/components/NewData.js:
```diff
@@ -0,0 +1,104 @@
+          <div>{n.content.split('<p>').join(' ').split('&#x27;').join('`')}</div>
+          <div>{n.time_ago}</div>
+          <button onClick={()=> setIsHidden(state => !state)}>{isHidden ? 'v':'^'}</button>
+          <ul>{n.comments !==[] && !isHidden ?  <Comments items={n.comments} /> : null}</ul>
```
`n.comments !==[]` неправильная проверка, массив ссылочный тип, при логическом сравнении массива сравниваются ссылки, а не содержимое.
Так как пустой массив ты создаешь здесь то у него всегда будет новая ссылка и она всегда будет отличаться от ссылки `n.comments`
Хорошая статья - https://doka.guide/js/ref-type-vs-value-type/
Бтв советую на этом сайте тоже почитывать интересные моменты

Stilwuc commented on hackers-site/src/components/NewData.js:
```diff
@@ -0,0 +1,104 @@
+              <Wrapp>
+                <div>by &nbsp;{data.user}</div>
+                <div>&nbsp;{data.time_ago}</div>
+                <div>&nbsp;{data.comments_count}&nbsp;comments</div>
```
Лучше пробелы делать так
```suggestion
                <div>{` ${data.comments_count} comments`}</div>
```

Stilwuc commented on hackers-site/src/components/New.js:
```diff
@@ -0,0 +1,211 @@
+      
+  
+    const refreshPage = () => {
+       refetch();
```
Обновляешь только первые 30 элементов, а нужно все

anmivin replied:
фиксед

Stilwuc reviewed (changes requested):
Дизайн прикольный концептуальный
По коду пару косяков есть поправь как будет время
Но так в целом гуд гуд

anmivin committed:
удалено всё ненужное
переименованы компоненты
useQuery запросы и вся функция New() соответственно изменены, выводится нужное количество элементов

anmivin committed:
выбранная новость теперь отображается и после перезагрузки
пофикшена рекурсия
правильная обработкак html фрагментов

Stilwuc commented on hackers-site/src/index.js:
```diff
@@ -0,0 +1,27 @@
+
+
+
+
```
А тут prettier'ом не прошлась)
Советую просто запустить команду `prettier --write .`

Stilwuc commented on hackers-site/src/NewsItemContext.js:
```diff
@@ -0,0 +1,5 @@
+
+const NewsItemContext = createContext();
+
+export default NewsItemContext;
```
Вроде же теперь не юзаешь контекст
Можно выпилить значит

Stilwuc commented on hackers-site/src/App.js:
```diff
@@ -0,0 +1,41 @@
+const queryClient = new QueryClient();
+function App() {
+  return (
+    <>
```
Тут и без фрагментов ок будет

anmivin committed:
заметила, что название функции непреднамеренно изменилось, исправила

Stilwuc commented on hackers-site/src/components/CommentTree.js:
```diff
@@ -0,0 +1,68 @@
+const TreeNode = ({ node }) => {
+  const [childVisible, setChildVisible] = useState(false);
+
+  const hasChild = node.comments.length !== 0 ? true : false;
```
Можно сократить
```suggestion
  const hasChild = !!node.comments.length;
```

Stilwuc commented on hackers-site/src/components/New.js:
```diff
@@ -0,0 +1,83 @@
+  };
+
+  const goToNew = (key) => {
+    navigate('/news', { state: { param: key } });
```
state роутера живет тоже до f5)
Вот что будет когда обновлю
![image](https://user-images.githubusercontent.com/44724024/185760285-8a3c6300-1c98-49b6-8383-9b8f1a831b29.png)

Stilwuc commented on hackers-site/src/components/New.js:
```diff
@@ -0,0 +1,83 @@
+      const res3 = await axios.get('https://api.hnpwa.com/v0/newest/3.json');
+      const res4 = await axios.get('https://api.hnpwa.com/v0/newest/4.json');
+      const newsList = [...res1.data, ...res2.data, ...res3.data, ...res4.data];
+      return newsList.slice(1, 101);
```
А почему не с 0 до 100?)

Stilwuc reviewed (changes requested):
Еще парочку)

anmivin committed:
добавлен local storage
исправлены остальные недочёты

anmivin committed:
id новости передаётся с помощью useParams
